### DISCUSSION TITLE
Combobox filter behaviour can be improved.

### DISCUSSION THREAD

**Discussion by thevipinmishra** [chakra-ui/ark#2674] [General]
Okay, first of all, thanks a lot for creating a beautiful library of unstyled components, it's been a blessing to integrate it in our design system so far.

I was working on the Combobox component and noticed that the list is not filtered by default on input value.

Minimal repro - https://stackblitz.com/edit/vitejs-vite-g1vovc?file=src%2FApp.tsx

Now, I've seen a lot of other Combobox primitives, where filtering is there out of the box.

For example, [Headless UI's combobox](https://headlessui.com/react/combobox) and [React Aria Components](https://react-spectrum.adobe.com/react-aria/ComboBox.html) to name a couple.

Is it intented to manage the filteration manually? this can surely be improved in my humble opinion.

This behaviour is also mentioned in the [W3C specs](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/).

![image](https://github.com/user-attachments/assets/92e5beeb-6f02-4077-9ff3-da11a2f4bba3)

**Comment by segunadebayo** [comment]
Correct! This is a design decision. The filtering logic should be handled by default in the userland code.

Feel free to use https://github.com/chakra-ui/atlas/blob/9f0ca36760af158c2e3ca6dc9faeb5360734ce36/website/src/demos/combobox.demo.tsx as a guide.

Even in Headless UI, you can see the filtering is handled by the userland code

![CleanShot 2024-07-23 at 13 51 25@2x](https://github.com/user-attachments/assets/9554b717-b334-47b5-bef9-faa89f846e97)

**Comment by sunwrobert** [comment]
@segunadebayo I think the examples on https://ark-ui.com/react/docs/components/combobox should be a bit more clear. e.g. In the Demo, we see that the results are filtered, but on the sample code, there is no filtering so it makes developers assume that it's handled somewhere lower level. This is because, for example https://github.com/pacocoursey/cmdk does it implicitly. I assumed that it was the same here but doesn't seem to be. Would've just been nice for it to be clearer. Otherwise, thanks for the library!

**Comment by segunadebayo** [comment]
The docs have been updated with a clearer filtering pattern using `useListCollection` and `useFilter`:

```tsx
import { Combobox, useListCollection } from '@ark-ui/react/combobox'
import { useFilter } from '@ark-ui/react/locale'

const { contains } = useFilter({ sensitivity: 'base' })
const { collection, filter } = useListCollection({
  initialItems: items,
  filter: contains,
})

<Combobox.Root collection={collection} onInputValueChange={(e) => filter(e.inputValue)}>
  {/* ... */}
</Combobox.Root>
```
